tasks/tasks.go

```diff
@@ -0,0 +1,194 @@
+package tasks
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/utils"
+)
+
+// TaskConfig configures tasks that run periodically
+type TaskConfig struct {
+	Name      string   `json:"name"`
+	Args      []string `json:"command"`
+	Frequency string   `json:"frequency"`
+	Timeout   string   `json:"timeout,omitempty"`
+
+	freqDuration    time.Duration
+	timeoutDuration time.Duration
+	cmd             *exec.Cmd
+	quitChannel     chan int
+	ticker          *time.Ticker
+	logWriters      []io.WriteCloser
+}
+
+// Task a task that runs periodically until killed
+type Task interface {
+	Start() error
+	Stop()
+}
+
+func (t *TaskConfig) closeLogs() {
+	if t.logWriters == nil {
+		return
+	}
+	for _, w := range t.logWriters {
+		if err := w.Close(); err != nil {
+			log.Errorf("Unable to close log writer : %v", err)
+		}
+	}
+}
+
+// NewTasks parses json config into an array of Tasks
+func NewTasks(raw json.RawMessage) ([]Task, error) {
+	var tasks []Task
+	if raw == nil {
+		return tasks, nil
+	}
+	var configs []*TaskConfig
+	if err := json.Unmarshal(raw, &configs); err != nil {
+		return nil, fmt.Errorf("Task configuration error: %v", err)
+	}
+	for _, t := range configs {
+		if err := parseTask(t); err != nil {
+			return nil, err
+		}
+		tasks = append(tasks, t)
+	}
+	return tasks, nil
+}
+
+func parseTask(task *TaskConfig) error {
+	if task.Args == nil || len(task.Args) == 0 {
+		return fmt.Errorf("Task did not provide a command")
+	}
+	if task.Name == "" {
+		task.Name = strings.Join(task.Args, " ")
+	}
+	freq, err := time.ParseDuration(task.Frequency)
+	if err != nil {
+		return fmt.Errorf("Unable to parse frequency %s: %v", task.Frequency, err)
+	}
+	task.freqDuration = freq
+	task.timeoutDuration = freq
+	if task.Timeout != "" {
+		timeout, err := time.ParseDuration(task.Timeout)
+		if err != nil {
+			return fmt.Errorf("Unable to parse timeout %s: %v", task.Timeout, err)
+		}
+		task.timeoutDuration = timeout
+	} else {
+		task.Timeout = task.Frequency
+	}
+	task.quitChannel = make(chan int)
+	return nil
+}
+
+// Start starts the task
+func (t *TaskConfig) Start() error {
+
+	t.ticker = time.NewTicker(t.freqDuration)
+	quit := make(chan int)
+
+	// Accept quit signal
+	go func() {
+		select {
+		case <-t.quitChannel:
+			log.Debugf("task[%s].Start#gofunc <-quitChannel", t.Name)
+			t.ticker.Stop()
+			quit <- 0
+			t.kill(t.cmd)
+			return
+		}
+	}()
+
+	// Run command
```

Comment:
I *think* this is in its own goroutine because `t.execute()` should be blocking on wait, but `run` includes a handler for this same channel. Am I missing something here?